aya-log-common/src/lib.rs

```diff
@@ -105,55 +107,62 @@ mod userspace {
     unsafe impl aya::Pod for DisplayHint {}
 }
 
-struct TagLenValue<'a, T> {
-    tag: T,
-    value: &'a [u8],
+struct TagLenValue<T, V> {
```

Comment:
This code is used with both arrays and slices. In the array case (anything that uses `to_ne_bytes`), the array is already on the stack, so I don't think this changes the representation. In the slice case, we now pass an iterator, which is possibly bigger than a slice header - but wouldn't you expect the compiler to optimize away the iterator?

Anyway to answer your question: this is necessary because it isn't otherwise possible to log a byte array as hex (apart from adding intrinsic support for it, as y'all did for MAC addresses). What am I missing?

Reply:
> This code is used with both arrays and slices. In the array case (anything that uses `to_ne_bytes`), the array is already on the stack, so I don't think this changes the representation. In the slice case, we now pass an iterator, which is possibly bigger than a slice header - but wouldn't you expect the compiler to optimize away the iterator?

I haven't done a proper review yet, I didn't realize you're passing it an iterator. I do expect the iterator to be optimized away.

> Anyway to answer your question: this is necessary because it isn't otherwise possible to log a byte array as hex (apart from adding intrinsic support for it, as y'all did for MAC addresses). What am I missing?

I think we should add intrinsic support for it. We don't want to encourage people to do any kind of formatting on the ebpf side, especially for a potentially large buffer => hex conversion. In a previous iteration of aya-log we did do formatting on the ebpf side, and it didn't work for several reasons. The verifier is not good at tracking loop bounds and having branches inside a loop quickly exhausts the instruction budget. This is why the new approach is "somehow create a flat byte array from the thing to be formatted and do the formatting in userspace". This isn't a thing we invented btw, logging crates for smol embedded hardware work like this too.

Reply:
I should add that i'm not against what this patch does tho. I think we might still want to merge this change as using iterators instead of copy_from_slice might actually end up in fewer copies. I'll do a proper review later today.

Reply:
Well, we can back out the last commit and add intrinsic support instead. Happy to do whatever you prefer.

Reply:
Yeah intrinsic support would be great :)

Reply:
Done!